tag_generator/src/lib.rs

```diff
@@ -0,0 +1,92 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use imap_types::core::Tag;
+#[cfg(not(debug_assertions))]
+use rand::distributions::{Alphanumeric, DistString};
+
+static GLOBAL_TAG_GENERATOR_COUNT: AtomicUsize = AtomicUsize::new(1);
+
+#[derive(Debug)]
+pub struct TagGeneratorGenerator {}
+
+impl TagGeneratorGenerator {
+    /// Generate an instance of a `TagGenerator`
+    ///
+    /// Returns a `TagGenerator` generating tags with a unique prefix.
+    pub fn generate() -> TagGenerator {
+        let global = GLOBAL_TAG_GENERATOR_COUNT.fetch_add(1, Ordering::SeqCst);
+        let counter = 1;
+
+        TagGenerator { global, counter }
+    }
+}
+
+#[derive(Debug)]
+pub struct TagGenerator {
+    global: usize,
+    counter: usize,
+}
+
+impl TagGenerator {
+    /// Generate a unique `Tag`
+    ///
+    /// The tag has the form `<Instance>.<Counter>.<Random>`, and is guaranteed to be unique and not
+    /// guessable ("forward-secure").
+    ///
+    /// Rational: `Instance` and `Counter` improve IMAP trace readability.
+    /// The non-guessable `Random` hampers protocol-confusion attacks (to a limiting extend).
+    pub fn generate(&mut self) -> Tag<'static> {
+        #[cfg(not(debug_assertions))]
+        let token = Alphanumeric.sample_string(&mut rand::thread_rng(), 8);
+
+        // Minimize randomness during development
+        #[cfg(debug_assertions)]
+        let token = String::from("FIXED");
```

Comment:
The idea here is to faciliate security analyses and fuzz-testing. It's helpful to deactivate randomness by external means w/o the need to modify the code. It should probably be a feature flag, and I agree it could be surprising. Still, I would prefer to not block on this now. Let's open an issue to track this and revisit later.